src/samplers/multinomial.jl

```diff
@@ -53,7 +53,7 @@ function _rand!(rng::AbstractRNG, s::MultinomialSampler,
                 x::AbstractVector{<:Real})
     n = s.n
     k = length(s)
-    if n^2 > k
+    if k < n/50+2
```

Comment:
I guess this changed because of the different alg complexity of the algorithm used in AliasTables? Is this particular choice based on benchmarks?

Reply:
Yes, it is based on benchmarks. I think the change is due both to a different constant factor in the new alias tables and due to the old heuristic simply not being optimal. The latter reason being why I can report such huge performance improvements in multinomial sampling in the OP.

Reply:
diff to do heuristic tuning
```diff
diff --git a/src/samplers/multinomial.jl b/src/samplers/multinomial.jl
index 95df1cf6..781a0cb3 100644
--- a/src/samplers/multinomial.jl
+++ b/src/samplers/multinomial.jl
@@ -49,11 +49,18 @@ function MultinomialSampler(n::Int, prob::Vector{<:Real})
     return MultinomialSampler(n, prob, AliasTable(prob))
 end
 
+const LH_HEURISTIC = Ref{Bool}(true)
+const LH_PRINT = Ref{Bool}(true)
+function lh_heuristic(n, k)
+    LH_PRINT[] && @show n, k
+    LH_HEURISTIC[]
+end
+
 function _rand!(rng::AbstractRNG, s::MultinomialSampler,
                 x::AbstractVector{<:Real})
     n = s.n
     k = length(s)
-    if k < n/50+2
+    if lh_heuristic(n, k)
         multinom_rand!(rng, n, s.prob, x)
     else
         # Use an alias table
```

Annotated REPL session
```julia
julia> using Distributions, Chairmarks

julia> normalize(x) = x/sum(x)
normalize (generic function with 1 method)

julia> d = Distributions.Multinomial(17, normalize(rand(29)));

julia> rand(d, 1);
(n, k) = (17, 29)

julia> Distributions.LH_PRINT[] = false;

julia> Distributions.LH_HEURISTIC[] = true;

julia> function f(n, k)
           Distributions.LH_HEURISTIC[] = true
           display(@b Distributions.Multinomial(n, normalize(rand(k))) rand(_, 100))
           Distributions.LH_HEURISTIC[] = false
           display(@b Distributions.Multinomial(n, normalize(rand(k))) rand(_, 100))
       end
f (generic function with 1 method)

julia> f(1, 1000)
567.008 μs (8 allocs: 805.234 KiB)
48.542 μs (8 allocs: 805.234 KiB)

julia> f(1000, 1)
815.429 ns (3 allocs: 992 bytes)
183.503 μs (3 allocs: 992 bytes)

julia> # high n => true, low n => false.

julia> f(100, 100)
443.464 μs (7 allocs: 81.141 KiB)
16.792 μs (7 allocs: 81.141 KiB)

julia> # similar => false

julia> f(5000, 100)
815.845 μs (7 allocs: 81.141 KiB)
754.261 μs (7 allocs: 81.141 KiB)

julia> # n 50x bigger -> either way

julia> f(50000, 1000)
8.910 ms (8 allocs: 805.234 KiB)
7.556 ms (8 allocs: 805.234 KiB)

julia> f(500000, 10000)
80.750 ms (8 allocs: 7.956 MiB, 0.37% gc time)
82.586 ms (8 allocs: 7.956 MiB, 0.31% gc time)

julia> f(5000000, 100000)
878.673 ms (8 allocs: 79.057 MiB, 0.38% gc time, without a warmup)
861.426 ms (8 allocs: 79.057 MiB, 0.11% gc time, without a warmup)

julia> # scales up at a stable 50x.

julia> f(5000, 100)
1.288 ms (7 allocs: 81.141 KiB)
754.636 μs (7 allocs: 81.141 KiB)

julia> f(5000, 100)
784.178 μs (7 allocs: 81.141 KiB)
754.636 μs (7 allocs: 81.141 KiB)

julia> f(500, 10)
67.084 μs (6 allocs: 8.312 KiB)
76.543 μs (6 allocs: 8.312 KiB)

julia> f(50, 1)
1.304 μs (3 allocs: 992 bytes)
7.521 μs (3 allocs: 992 bytes)

julia> # Very small -> true

julia> f(50, 3)
9.959 μs (6 allocs: 2.625 KiB)
7.979 μs (6 allocs: 2.625 KiB)

julia> # n == 50(k-2) -> similar, but refactor to n/50+2 == k to avoid overflow on massive k. IDK if overflow is an issue and I dont want to have to worry about it nor do I want others reading this to have to worry about it. A single floating point multiplicatiob by the constant 1/50 should be cheap compared to these runtimes.
```

Reply:
please note i have a similar PR (and similar benchmarking) here https://github.com/JuliaStats/Distributions.jl/pull/1834

maybe the PRs can be orthogonalized this one only changes the alias table type and mine is left to change the multinomial sampler?

Reply:
Sure! 

I started benchmarking multinomial samplers to confirm that the performance improvements I saw translated to slightly downstream usages, and then when I realized that the heuristic was significantly suboptimal I re-tuned it. It's a very local optimization—feel free to overwrite it with a new, more well chosen heuristic, especially if you are also changing the costs of the branches.

My main contribution is [AliasTables.sample](https://aliastables.lilithhafner.com/dev/#AliasTables.sample) which maps UInt64 to an Int according to specified weights. I'm not totally familiar with nor in love with the Random API, so it's quite likely that you can make further improvements by leveraging that function more efficiently for a particular downstream usecase (e.g. bulk generation of UInt64 is 2-3x faster than scalar generation because it leverages the implementation details of the xoshiro rng. I don't take advantage of that in AliasTables)

Reply:
```suggestion
    if n^2 > k
```